repo/blob/pit/pit.go

```diff
@@ -0,0 +1,161 @@
+// Package pit provides a Point-in-time view of a versioned blob store
+package pit
+
+import (
+	"context"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/kopia/kopia/repo/blob"
+	"github.com/kopia/kopia/repo/blob/readonly"
+)
+
+type pitStorage struct {
+	blob.Storage
+
+	vs          VersionedStorage
+	pointInTime time.Time
+}
+
+func (s pitStorage) ListBlobs(ctx context.Context, blobIDPrefix blob.ID, cb func(bm blob.Metadata) error) error {
+	var (
+		previousID blob.ID
+		vs         []VersionMetadata
+	)
+
+	err := s.vs.ListBlobVersions(ctx, blobIDPrefix, func(vm VersionMetadata) error {
+		if vm.BlobID != previousID {
+			// different blob, process previous one
+			if v, found := newestAtUnlessDeleted(vs, s.pointInTime); found {
+				if err := cb(v.Metadata); err != nil {
+					return err
+				}
+			}
+
+			previousID = vm.BlobID
+			vs = vs[:0] // reset for next blob
+		}
+
+		vs = append(vs, vm)
+
+		return nil
+	})
+	if err != nil {
+		return errors.Wrapf(err, "could not list blob versions at time %s", s.pointInTime)
+	}
+
+	// process last blob
+	if v, found := newestAtUnlessDeleted(vs, s.pointInTime); found {
+		if err := cb(v.Metadata); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (s pitStorage) GetBlob(ctx context.Context, blobID blob.ID, offset, length int64) ([]byte, error) {
+	// getMetadata returns the specific blob version at time t
+	m, err := s.getMetadata(ctx, blobID)
+	if err != nil {
+		return nil, err
+	}
+
+	b, err := s.vs.GetBlobWithVersion(ctx, blobID, m.Version, offset, length)
+
+	return b, errors.Wrapf(err, "could not get blob: %s, version: %s", blobID, m.Version)
+}
+
+func (s pitStorage) GetMetadata(ctx context.Context, blobID blob.ID) (blob.Metadata, error) {
+	m, err := s.getMetadata(ctx, blobID)
+
+	return m.Metadata, err
+}
+
+func (s pitStorage) getMetadata(ctx context.Context, blobID blob.ID) (VersionMetadata, error) {
+	var vml []VersionMetadata
+
+	if err := s.vs.GetBlobVersions(ctx, blobID, func(m VersionMetadata) error {
+		// only include versions older than s.pointInTime
+		if !m.Timestamp.After(s.pointInTime) {
+			vml = append(vml, m)
+		}
+
+		return nil
+	}); err != nil {
+		return VersionMetadata{}, errors.Wrapf(err, "could not get version metadata for blob %s", blobID)
+	}
+
+	if v, found := newestAtUnlessDeleted(vml, s.pointInTime); found {
+		return v, nil
+	}
+
+	return VersionMetadata{}, blob.ErrBlobNotFound
+}
+
+func newestAtUnlessDeleted(vs []VersionMetadata, t time.Time) (v VersionMetadata, found bool) {
+	vs = getOlderThan(vs, t)
```

Comment:
This `getOlderThan` call seems unneeded when `newestAtUnlessDeleted` is called in `getMetadata`, yeah? At this point all versions in `v` are already earlier than `s.pointInTime` after the same filter in the `GetBlobVersions` callback.

At first it seemed like redundant work, but on rereading, it should always return in the first loop of `getOlderThan`, so doesn't add much overhead